File: src/handoff.js
// Signing in through the player's browser and taking the answer back (handoff.rs).
//
// Apple and Google cannot be signed in with here. Google refuses its own pages inside an
// embedded web view, and Apple has no sheet to open on Windows; and either page opened in
// the player's browser answers there, where the cookies are not this window's. So the site
// hands the sign-in out and takes it back: this asks the site to start a handoff, the app
// opens the browser at the URL it gets, and this asks the site until the browser has
// finished -- at which point the site signs *this* window in (src/handoff.rs in
// oeee-cafe/web).
//
// Every request is made here, in the page, so each carries the page's cookie and origin.
// The app's whole part is opening a browser: it never holds the session, and never sees
// the secret that claims the handoff.
//
// Defined in every page the window loads, as the bridge is, and for the same reason: the
// window only ever holds the site and the bundled loader, and a page Tauri has given no
// IPC to is left without this rather than given one that cannot carry.
(function () {
  if (!window.__TAURI_INTERNALS__) return;

  // Asks the app to open `url` in the browser. The app checks it is the site's own.
  function openInBrowser(url) {
    window.__TAURI_INTERNALS__
      .invoke("plugin:event|emit", { event: "oeee-handoff", payload: url })
      .catch(function () {});
  }

  // The handoff under way: its id, its secret, and where it is going afterwards.
  var pending = null;
  var asking = null;
  // Whether an ask is in flight. Two at once can spend the handoff between them: one gets
  // the sign-in and the other gets "there is no such handoff", and whichever lands second
  // decides what the page does.
  var inFlight = false;

  // How often to ask whether the browser has finished, and how long to keep asking. The
  // site forgets a handoff after fifteen minutes, so there is nothing to find after that.
  var ASK_EVERY = 2000;
  var GIVE_UP_AFTER = 15 * 60 * 1000;

  function stop() {
    if (asking) {
      clearInterval(asking);
      asking = null;
    }
    pending = null;
    inFlight = false;
  }

  function form(fields) {
    var body = new URLSearchParams();
    for (var name in fields) {
      if (fields[name]) body.set(name, fields[name]);
    }
    return body.toString();
  }

  function post(path, fields) {
    return fetch(path, {
      method: "POST",
      credentials: "same-origin",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: form(fields)
    }).then(function (response) {
      return response.ok ? response.json() : null;
    });
  }

  // Asks the site whether the browser has finished, and acts on what it says.
  function ask(confirmed) {
    if (!pending || inFlight) return;
    inFlight = true;
    var fields = { id: pending.id, secret: pending.secret };
    if (confirmed) fields.confirm = "1";
    post("/auth/handoff/claim", fields)
      .then(function (answer) {
        inFlight = false;
        // The site could not answer: nothing has been decided, so keep asking rather than
        // giving up on a sign-in that may well have taken.
        if (!answer || !pending) return;
        if (answer.status === "waiting" || answer.status === "failed") return;
        if (answer.status === "confirm") {
          // Linking the browser's account to the one signed in here. The site words the
          // question, because it is the one that knows the reader's language; the app
          // draws it as its own dialog (dialogs.rs).
          if (window.confirm(answer.message)) {
            ask(true);
          } else {
            stop();
          }
          return;
        }
        if (answer.status === "ready") {
          var next = answer.next || "/";
          stop();
          // Replaced, not pushed: the page signed in from is not left behind the one it
          // lands on, so Back does not return to a sign-in form for an account already
          // signed in.
          location.replace(next);
          return;
        }
        // "unknown", or anything a later site says that this does not know.
        stop();
      })
      .catch(function () {
        inFlight = false;
      });
  }

  window.oeeeHandoffAuth = {
    // Starts a sign-in with `provider` in the browser, going on to `next` afterwards.
    // The app calls this when it stops the link to /auth/<provider>.
    begin: function (provider, next) {
      if (pending) return;
      post("/auth/handoff/start", { provider: provider, next: next })
        .then(function (started) {
          if (!started || !started.id || !started.secret || !started.url) {
            stop();
            return;
          }
          pending = started;
          openInBrowser(started.url);
          asking = setInterval(function () {
            ask(false);
          }, ASK_EVERY);
          setTimeout(function () {
            if (pending) stop();
          }, GIVE_UP_AFTER);
        })
        .catch(stop);
    },

    // The window is in front again, so somebody may have just come back from the browser:
    // ask now rather than waiting for the next turn of the clock.
    resume: function () {
      if (pending) ask(false);
    }
  };
})();

File: src/handoff.rs
//! Signing in with Apple or Google, which this window cannot do itself.
//!
//! Google refuses its own sign-in pages inside an embedded web view, and
//! Apple ships no sheet to open on Windows. Opening either in the player's
//! browser signs them in *there*, where the cookies are not this window's, so
//! the site hands the sign-in out and takes it back (src/handoff.rs in
//! oeee-cafe/web): the page asks the site to start a handoff, the app opens
//! the browser at the URL it is given, and the page asks the site until the
//! browser has finished, at which point the site signs this window in.
//!
//! The page's half is the site's own (app_sign_in.jinja in oeee-cafe/web),
//! shared with the other apps, and it starts itself: the page takes the
//! press on the sign-in buttons, so the app never sees the link. The app's
//! whole part is opening a browser when the page asks (`browse` on the
//! bridge, bridge.rs), and saying when it could not or when the window comes
//! back. Every request is the page's, so each carries the page's cookie and
//! origin; the app never holds the session and never sees the secret that
//! claims the handoff.

use tauri::{AppHandle, Manager};
use tauri_plugin_opener::OpenerExt;
use url::Url;

use crate::{site, WINDOW};

/// Opens the browser where the page asks, once it has a handoff to carry
/// (`browse` on the bridge), or tells the page it could not.
///
/// Only at the site's own URLs: the page is the site's, but the window is the
/// app's, and a page is not given the run of whatever the browser will open.
pub fn browse(app: &AppHandle, site: &Url, url: &str) {
    let opened = match Url::parse(url) {
        Ok(url) if site::is_site(&url, site) => app
            .opener()
            .open_url(url.as_str(), None::<&str>)
            .map_err(|error| eprintln!("could not open the sign-in in the browser: {error}"))
            .is_ok(),
        _ => {
            eprintln!("a page asked for a browser somewhere that is not the site: {url}");
            false
        }
    };
    if !opened {
        if let Some(window) = app.get_webview_window(WINDOW) {
            let _ = window.eval("window.oeeeApp && window.oeeeApp.signIn && window.oeeeApp.signIn.unopened();");
        }
    }
}

/// The window coming back to the front is somebody returning from the
/// browser, most likely having just signed in: the page asks the site then
/// and there rather than waiting for the next turn of its own clock.
pub fn on_window_event<R: tauri::Runtime>(window: &tauri::Window<R>, event: &tauri::WindowEvent) {
    if !matches!(event, tauri::WindowEvent::Focused(true)) {
        return;
    }
    if let Some(window) = window.get_webview_window(WINDOW) {
        ask_now(&window);
    }
}

fn ask_now<R: tauri::Runtime>(window: &tauri::WebviewWindow<R>) {
    let _ = window.eval("window.oeeeApp && window.oeeeApp.signIn && window.oeeeApp.signIn.resume();");
}

/// The scheme the browser is sent to once the provider has answered.
///
/// Registered so the browser can knock: the page in the window is asking the
/// site every couple of seconds anyway, and would get there on its own the
/// moment somebody switched back, but nobody should have to go looking for
/// the window. It carries nothing -- the window is already holding the id
/// and the secret that claim the sign-in -- so an app that took the scheme
/// for itself would learn nothing by it.
pub const RETURN_SCHEME: &str = "oeee-cafe";

/// Hears the browser knock while the app is already running.
///
/// Where a scheme arrives as a second launch instead, single-instance hands
/// it over and calls [`returned`] the same way (main.rs); both are wired
/// because which one happens is the platform's business, not ours.
pub fn listen_for_return(app: &AppHandle) {
    use tauri_plugin_deep_link::DeepLinkExt;
    // Windows wants the scheme registered with the system, and the two
    // Windows builds want it registered differently.
    //
    // Steam ships loose files into a depot folder and installs no package, so
    // there is nothing to declare it for us and this is what registers it.
    // The Store build is an MSIX, which takes its protocols from its manifest
    // and only from there (msstore/AppxManifest.xml); inside the package this
    // registry write is virtualised and does nothing, which is harmless.
    //
    // A failure is not worth stopping for either way: all it costs is the
    // knock, and the page still asks the site every couple of seconds.
    #[cfg(windows)]
    if let Err(error) = app.deep_link().register(RETURN_SCHEME) {
        eprintln!("could not register {RETURN_SCHEME}://: {error}");
    }
    let app = app.clone();
    app.clone().deep_link().on_open_url(move |_event| {
        returned(&app);
    });
}

/// Brings the window forward when the browser knocks, and asks the site at
/// once rather than waiting for the page's own clock.
pub fn returned(app: &AppHandle) {
    let Some(window) = app.get_webview_window(WINDOW) else {
        return;
    };
    let _ = window.unminimize();
    let _ = window.show();
    let _ = window.set_focus();
    ask_now(&window);
}
